// @funeste38/bat/dist/bat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bat = void 0;
class Bat {
    constructor(opts = {}) {
        this.counter = 0;
        this.inFlight = new Map();
        this.baselineMs = opts.baselineMs ?? 500;
        this.alpha = opts.alpha ?? 0.2;
    }
    start(key) {
        const id = `${key}:${++this.counter}`;
        this.inFlight.set(id, { startedAt: Date.now(), attempts: 1 });
        return id;
    }
    stop(id) {
        const rec = this.inFlight.get(id);
        if (!rec)
            return null;
        const rtt = Date.now() - rec.startedAt;
        this.baselineMs = (1 - this.alpha) * this.baselineMs + this.alpha * rtt;
        this.inFlight.delete(id);
        return { rtt, cls: this.classify(rtt), attempts: rec.attempts };
    }
    classify(rtt) {
        if (rtt < this.baselineMs * 0.5)
            return 'short';
        if (rtt < this.baselineMs * 1.5)
            return 'normal';
        if (rtt < this.baselineMs * 3)
            return 'slow';
        return 'timeout';
    }
    markRetry(id) {
        const rec = this.inFlight.get(id);
        if (!rec)
            return null;
        rec.attempts += 1;
        rec.startedAt = Date.now();
        return rec;
    }
    listInFlight() { return [...this.inFlight.entries()]; }
}
exports.Bat = Bat;
