// @funeste38/bat/dist/batWings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatWings = void 0;
const profiles_1 = require("./profiles");
class BatWings {
    constructor() {
        this.state = {
            rate: 1,
            retries: 0,
            safeMode: false
        };
        this.consecutiveTimeouts = 0;
        this.maxTimeouts = 3;
        this.lastAdjustAt = Date.now();
        this.profile = profiles_1.defaultProfiles.chat;
    }
    selectProfile(type) {
        this.profile = profiles_1.defaultProfiles[type] ?? profiles_1.defaultProfiles.chat;
        return this.profile;
    }
    applyProfile(echoClass) {
        const profile = this.profile;
        // small cooldown to avoid thrashing
        const now = Date.now();
        if (now - this.lastAdjustAt < 50)
            return;
        this.lastAdjustAt = now;
        if (echoClass === 'short')
            this.state.rate = Math.min(3, this.state.rate + profile.rateGain);
        if (echoClass === 'normal')
            this.state.rate = this.state.rate; // stable
        if (echoClass === 'slow')
            this.state.rate = Math.max(0.2, this.state.rate + profile.rateLoss);
        if (echoClass === 'timeout') {
            this.state.rate = Math.max(0.1, this.state.rate + profile.rateLoss * 3);
            this.consecutiveTimeouts++;
        }
        // clamp
        this.state.rate = Math.max(0.2, Math.min(3, this.state.rate));
    }
    adjustRate(cls) {
        // legacy fallback to keep compatibility
        this.applyProfile(cls);
    }
    computeTimeout(baseline) {
        return Math.max(50, Math.round(baseline * (this.profile.timeoutFactor ?? 1)));
    }
    shouldRetry(attempts) {
        return attempts < (this.profile.maxRetries ?? 0);
    }
    shouldAbort() {
        return this.consecutiveTimeouts >= this.maxTimeouts || this.state.safeMode;
    }
    resetTimeouts() {
        this.consecutiveTimeouts = 0;
    }
    drop(id) {
        // placeholder for drop action: in a real system we'd cancel network requests
        // here we just log
        // eslint-disable-next-line no-console
        console.warn(`BatWings.drop: dropping ${id}`);
    }
    async emergencyRecovery() {
        // simple unsync recovery: wait, then reset rate and timeouts
        // eslint-disable-next-line no-console
        console.warn('BatWings: emergency recovery — pausing briefly');
        await new Promise((r) => setTimeout(r, 300));
        this.state.rate = 1;
        this.resetTimeouts();
        // eslint-disable-next-line no-console
        console.warn('BatWings: recovery complete');
    }
}
exports.BatWings = BatWings;
