// @funeste38/bat/dist/instinct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Instinct = void 0;
class Instinct {
    constructor(opts = {}) {
        this.defaults = opts.defaults ?? {};
        this.config = opts.config ?? {};
        this.envPriority = opts.envPriority ?? ['env', 'config', 'defaults'];
    }
    detectSlots(cmd) {
        const re = /\{([A-Z0-9_\-]+)\}/g;
        const out = [];
        let m;
        while ((m = re.exec(cmd)) !== null) {
            out.push(m[1]);
        }
        return Array.from(new Set(out));
    }
    resolveSlot(name) {
        for (const src of this.envPriority) {
            if (src === 'env' && process.env[name])
                return process.env[name];
            if (src === 'config' && this.config[name])
                return this.config[name];
            if (src === 'defaults' && this.defaults[name])
                return this.defaults[name];
        }
        return undefined;
    }
    fill(cmd) {
        const slots = this.detectSlots(cmd);
        const missing = [];
        let filled = cmd;
        for (const s of slots) {
            const val = this.resolveSlot(s);
            if (val === undefined) {
                missing.push(s);
                continue;
            }
            const re = new RegExp(`\\{${s}\\}`, 'g');
            filled = filled.replace(re, val);
        }
        return { cmd: filled, missing };
    }
}
exports.Instinct = Instinct;
