// @funeste38/bat/dist/wings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wings = void 0;
const profiles_1 = require("./profiles");
class Wings {
    constructor() {
        this.state = { rate: 1, retries: 0, safeMode: false };
        this.consecutiveTimeouts = 0;
        this.maxTimeouts = 3;
        this.lastAdjustAt = Date.now();
        this.profile = profiles_1.defaultProfiles.chat;
    }
    selectProfile(type) { this.profile = profiles_1.defaultProfiles[type] ?? profiles_1.defaultProfiles.chat; return this.profile; }
    applyProfile(echoClass) {
        const profile = this.profile;
        const now = Date.now();
        if (now - this.lastAdjustAt < 50)
            return;
        this.lastAdjustAt = now;
        if (echoClass === 'short')
            this.state.rate = Math.min(3, this.state.rate + profile.rateGain);
        if (echoClass === 'normal')
            this.state.rate = this.state.rate;
        if (echoClass === 'slow')
            this.state.rate = Math.max(0.2, this.state.rate + profile.rateLoss);
        if (echoClass === 'timeout') {
            this.state.rate = Math.max(0.1, this.state.rate + profile.rateLoss * 3);
            this.consecutiveTimeouts++;
        }
        this.state.rate = Math.max(0.2, Math.min(3, this.state.rate));
    }
    computeTimeout(baseline) { return Math.max(50, Math.round(baseline * (this.profile.timeoutFactor ?? 1))); }
    shouldRetry(attempts) { return attempts < (this.profile.maxRetries ?? 0); }
    shouldAbort() { return this.consecutiveTimeouts >= this.maxTimeouts || this.state.safeMode; }
    resetTimeouts() { this.consecutiveTimeouts = 0; }
    reset() { this.state.rate = 1; this.resetTimeouts(); this.state.retries = 0; this.state.safeMode = false; }
}
exports.Wings = Wings;
