// @funeste38/rome/dist/config.js
import fs from "fs";
import path from "path";
import { globalLogger } from "./logger.js";
/** Charge la configuration rome.json depuis la racine du projet */
export function loadConfig(root = process.cwd()) {
    const configPath = path.join(root, 'rome.json');
    if (!fs.existsSync(configPath))
        return null;
    try {
        const content = fs.readFileSync(configPath, 'utf8');
        return JSON.parse(content);
    }
    catch (e) {
        globalLogger.warning('Erreur de lecture du fichier rome.json:', e);
        return null;
    }
}
/** Parse les options d'affichage depuis les arguments CLI */
export function parseOutputOptions(args) {
    const options = {};
    // Parse les flags
    for (const arg of args) {
        switch (arg) {
            case '--plain':
                options.plain = true;
                break;
            case '--json':
                options.json = true;
                break;
            case '--no-emoji':
                options.noEmoji = true;
                break;
            case '--ci':
                options.ci = true;
                break;
        }
    }
    return options;
}
/** Fusionne la config utilisateur avec les valeurs par défaut */
export function mergeConfig(userConfig) {
    const defaults = {
        workspaces: {
            patterns: [
                "server", "backend", "api", "web", "front", "client", "app",
                "apps/*", "packages/*", "services/*", "src/*"
            ],
            exclude: ["node_modules", ".git", "dist", "build"]
        },
        packageManager: 'auto',
        scripts: {
            build: 'build',
            test: 'test',
            lint: 'lint',
            dev: 'dev'
        },
        output: {
            plain: false,
            json: false,
            noEmoji: false,
            ci: false
        }
    };
    if (!userConfig)
        return defaults;
    return {
        workspaces: {
            patterns: userConfig.workspaces?.patterns || defaults.workspaces.patterns,
            exclude: userConfig.workspaces?.exclude || defaults.workspaces.exclude
        },
        packageManager: userConfig.packageManager || defaults.packageManager,
        scripts: {
            ...defaults.scripts,
            ...userConfig.scripts
        },
        output: {
            ...defaults.output,
            ...userConfig.output
        }
    };
}
/** Fusionne les options CLI avec la config (CLI prioritaire) */
export function resolveOutputOptions(cliOptions, configOptions) {
    const defaults = {
        plain: false,
        json: false,
        noEmoji: false,
        ci: false
    };
    const config = configOptions || {};
    // CLI a la priorité, puis config, puis defaults
    const resolved = {
        plain: cliOptions.plain !== undefined ? cliOptions.plain : (config.plain || defaults.plain),
        json: cliOptions.json !== undefined ? cliOptions.json : (config.json || defaults.json),
        noEmoji: cliOptions.noEmoji !== undefined ? cliOptions.noEmoji : (config.noEmoji || defaults.noEmoji),
        ci: cliOptions.ci !== undefined ? cliOptions.ci : (config.ci || defaults.ci)
    };
    // --ci équivaut à --plain --no-emoji (même si spécifié via config)
    if (resolved.ci) {
        resolved.plain = true;
        resolved.noEmoji = true;
    }
    return resolved;
}
