// @funeste38/rome/dist/logger.js
import { format } from "util";
/** Format timestamp pour les logs */
function getTimestamp() {
    return new Date().toISOString().slice(11, 23); // HH:MM:SS.mmm
}
/** Transforme les emojis selon les options */
function transformEmojis(message, options) {
    if (options.noEmoji || options.plain) {
        // Supprime les emojis
        return message.replace(/[\u{1F600}-\u{1F64F}]|[\u{1F300}-\u{1F5FF}]|[\u{1F680}-\u{1F6FF}]|[\u{1F1E0}-\u{1F1FF}]|[\u{2600}-\u{26FF}]|[\u{2700}-\u{27BF}]/gu, '').trim();
    }
    return message;
}
/** Transforme le texte romain selon les options */
function transformRomanText(message, options) {
    if (options.plain) {
        // Remplace les termes romains par des équivalents simples
        return message
            .replace(/🏛️/g, 'Rome')
            .replace(/⚔️/g, 'Rome')
            .replace(/🏰/g, 'Rome')
            .replace(/👨‍✈️/g, 'Rome')
            .replace(/🏹/g, 'Rome')
            .replace(/🏗️/g, 'Rome')
            .replace(/⚔️/g, 'Rome')
            .replace(/🐎/g, 'Rome')
            .replace(/👑/g, 'Rome')
            .replace(/🦅/g, 'Rome')
            .replace(/🔫/g, 'Rome')
            .replace(/ARMÉE ROMAINE/g, 'Workspaces')
            .replace(/Formation/g, 'Workspace')
            .replace(/Capitaine/g, 'Folder')
            .replace(/Archer/g, 'Docs')
            .replace(/Catapulte/g, 'Scripts')
            .replace(/Infanterie/g, 'Code')
            .replace(/Cavalerie/g, 'Tests')
            .replace(/Centurion/g, 'Config')
            .replace(/Aquila/g, 'Main')
            .replace(/Balliste/g, 'Assets')
            .replace(/Force:/g, 'Files:')
            .replace(/Bataillon/g, 'Small')
            .replace(/Légion/g, 'Medium')
            .replace(/Armée/g, 'Large');
    }
    return message;
}
/** Logger avec préfixe workspace + timestamp */
export class WorkspaceLogger {
    workspace;
    options;
    constructor(workspace, options) {
        this.workspace = workspace;
        this.options = options;
    }
    formatMessage(level, message, ...args) {
        const formatted = args.length > 0 ? format(message, ...args) : message;
        const transformed = transformEmojis(transformRomanText(formatted, this.options), this.options);
        if (this.options.json) {
            const logEntry = {
                timestamp: new Date().toISOString(),
                level: level.replace(/[^\w]/g, '').toLowerCase(),
                workspace: this.workspace,
                message: transformed
            };
            return JSON.stringify(logEntry);
        }
        const timestamp = getTimestamp();
        const prefix = `[${timestamp}] [${this.workspace}] ${level}`;
        return `${prefix} ${transformed}`;
    }
    info(message, ...args) {
        console.log(this.formatMessage("ℹ️", message, ...args));
    }
    success(message, ...args) {
        console.log(this.formatMessage("✅", message, ...args));
    }
    warning(message, ...args) {
        console.warn(this.formatMessage("⚠️", message, ...args));
    }
    error(message, ...args) {
        console.error(this.formatMessage("❌", message, ...args));
    }
    debug(message, ...args) {
        if (process.env.DEBUG) {
            console.log(this.formatMessage("🐛", message, ...args));
        }
    }
}
/** Logger global pour les opérations générales */
export function createGlobalLogger(options) {
    return {
        info: (message, ...args) => {
            const formatted = args.length > 0 ? format(message, ...args) : message;
            const transformed = transformEmojis(transformRomanText(formatted, options), options);
            if (options.json) {
                const logEntry = {
                    timestamp: new Date().toISOString(),
                    level: 'info',
                    message: transformed
                };
                console.log(JSON.stringify(logEntry));
                return;
            }
            const timestamp = getTimestamp();
            console.log(`[${timestamp}] ℹ️ ${transformed}`);
        },
        success: (message, ...args) => {
            const formatted = args.length > 0 ? format(message, ...args) : message;
            const transformed = transformEmojis(transformRomanText(formatted, options), options);
            if (options.json) {
                const logEntry = {
                    timestamp: new Date().toISOString(),
                    level: 'success',
                    message: transformed
                };
                console.log(JSON.stringify(logEntry));
                return;
            }
            const timestamp = getTimestamp();
            console.log(`[${timestamp}] ✅ ${transformed}`);
        },
        warning: (message, ...args) => {
            const formatted = args.length > 0 ? format(message, ...args) : message;
            const transformed = transformEmojis(transformRomanText(formatted, options), options);
            if (options.json) {
                const logEntry = {
                    timestamp: new Date().toISOString(),
                    level: 'warning',
                    message: transformed
                };
                console.warn(JSON.stringify(logEntry));
                return;
            }
            const timestamp = getTimestamp();
            console.warn(`[${timestamp}] ⚠️ ${transformed}`);
        },
        error: (message, ...args) => {
            const formatted = args.length > 0 ? format(message, ...args) : message;
            const transformed = transformEmojis(transformRomanText(formatted, options), options);
            if (options.json) {
                const logEntry = {
                    timestamp: new Date().toISOString(),
                    level: 'error',
                    message: transformed
                };
                console.error(JSON.stringify(logEntry));
                return;
            }
            const timestamp = getTimestamp();
            console.error(`[${timestamp}] ❌ ${transformed}`);
        },
        debug: (message, ...args) => {
            if (process.env.DEBUG) {
                const formatted = args.length > 0 ? format(message, ...args) : message;
                const transformed = transformEmojis(transformRomanText(formatted, options), options);
                if (options.json) {
                    const logEntry = {
                        timestamp: new Date().toISOString(),
                        level: 'debug',
                        message: transformed
                    };
                    console.log(JSON.stringify(logEntry));
                    return;
                }
                const timestamp = getTimestamp();
                console.log(`[${timestamp}] 🐛 ${transformed}`);
            }
        }
    };
}
/** Logger global par défaut (legacy) */
export const globalLogger = createGlobalLogger({ plain: false, json: false, noEmoji: false, ci: false });
